File: ios/RNIceLink/Opus/OpusCodec.m
#import "OpusCodec.h"

@implementation OpusCodec

@synthesize percentLossToTriggerFEC=_percentLossToTriggerFEC;
@synthesize disableFEC=_disableFEC;

- (bool)fecActive
{
    return _fecActive;
}

- (id)init
{
    self = [super initWithPacketTime:20];
    if (self)
    {
        _disableFEC = NO;
        _percentLossToTriggerFEC = 5;
        _minimumReportsBeforeFEC = 2;
        
        _currentRTPSequenceNumber = -1;
        _lastRTPSequenceNumber = -1;
        
        _padep = [[FMIceLinkWebRTCBasicAudioPadep alloc] init];
    }
    return self;
}

- (NSMutableData *)encodeWithFrame:(FMIceLinkWebRTCAudioBuffer *)frame
{
    if (!_encoder)
    {
        _encoder = [[CocoaOpusEncoder alloc] initWithClockRate:self.clockRate channels:self.channels packetTime:self.packetTime];
        _encoder.quality = 0.5;
        _encoder.bitrate = 125;
    }
    
    CocoaOpusBuffer *buffer = [[CocoaOpusBuffer alloc] init];
    buffer.data = frame.data;
    buffer.index = frame.index;
    buffer.length = frame.length;
    return [_encoder encodeBuffer:buffer];
}

- (FMIceLinkWebRTCAudioBuffer *)decodeWithEncodedFrame:(NSMutableData *)encodedFrame
{
    if (!_decoder)
    {
        _decoder = [[CocoaOpusDecoder alloc] initWithClockRate:self.clockRate channels:self.channels packetTime:self.packetTime];
        [self.link getRemoteStream].disablePLC = YES;
    }
    
    if (_lastRTPSequenceNumber == -1)
    {
        _lastRTPSequenceNumber = _currentRTPSequenceNumber;
        return [self decodeNormalWithEncodedFrame:encodedFrame];
    }
    else
    {
        int sequenceNumberDelta = [FMIceLinkRTPPacket getSequenceNumberDeltaWithSequenceNumber:_currentRTPSequenceNumber lastSequenceNumber:_lastRTPSequenceNumber];
        if (sequenceNumberDelta <= 0)
        {
            return nil;
        }
        
        _lastRTPSequenceNumber = _currentRTPSequenceNumber;
        
        int missingPacketCount = sequenceNumberDelta - 1;
        NSMutableArray *previousFrames = [NSMutableArray arrayWithCapacity:missingPacketCount];
        
        int plcFrameCount = (missingPacketCount > 1) ? missingPacketCount - 1 : 0;
        if (plcFrameCount > 0)
        {
            [FMLog infoWithMessage:[NSString stringWithFormat:@"Adding %d frames of loss concealment to incoming audio stream. Packet sequence violated.", plcFrameCount]];
            for (int i = 0; i < plcFrameCount; i++)
            {
                [previousFrames addObject:[self decodePLC]];
            }
        }
        
        int fecFrameCount = (missingPacketCount > 0) ? 1 : 0;
        if (fecFrameCount > 0)
        {
            FMIceLinkWebRTCAudioBuffer *fecFrame = [self decodeFECWithEncodedFrame:encodedFrame];
            if (fecFrame == nil)
            {
                [previousFrames addObject:[self decodePLC]];
            }
            else
            {
                [previousFrames addObject:fecFrame];
            }
        }
        
        FMIceLinkWebRTCAudioBuffer *frame = [self decodeNormalWithEncodedFrame:encodedFrame];
        frame.previousBuffers = previousFrames;
        return frame;
    }
}

- (FMIceLinkWebRTCAudioBuffer *)decodePLC
{
    return [self decodeWithEncodedFrame:nil fec:NO];
}

- (FMIceLinkWebRTCAudioBuffer *)decodeFECWithEncodedFrame:(NSData *)encodedFrame
{
    return [self decodeWithEncodedFrame:encodedFrame fec:YES];
}

- (FMIceLinkWebRTCAudioBuffer *)decodeNormalWithEncodedFrame:(NSData *)encodedFrame
{
    return [self decodeWithEncodedFrame:encodedFrame fec:NO];
}

- (FMIceLinkWebRTCAudioBuffer *)decodeWithEncodedFrame:(NSData *)encodedFrame fec:(bool)fec
{
    CocoaOpusBuffer *buffer = [_decoder decodeFrame:encodedFrame fec:fec];
    if (buffer == nil)
    {
        return nil;
    }
    
    return [FMIceLinkWebRTCAudioBuffer audioBufferWithData:[NSMutableData dataWithBytesNoCopy:(void *)buffer.data.bytes length:buffer.data.length freeWhenDone:NO]
                                                     index:buffer.index
                                                    length:buffer.length];
}

- (NSMutableArray *)packetizeWithEncodedFrame:(NSMutableData *)encodedFrame
{
    return [_padep packetizeWithEncodedFrame:encodedFrame clockRate:self.clockRate packetTime:self.packetTime resetTimestamp:self.resetTimestamp];
}

- (NSMutableData *)depacketizeWithPacket:(FMIceLinkRTPPacket *)packet
{
    _currentRTPSequenceNumber = packet.sequenceNumber;
    
    return [_padep depacketizeWithPacket:packet];
}

- (void)processRTCPWithPackets:(NSMutableArray *)packets
{
    if (_encoder)
    {
        for (FMIceLinkRTCPPacket *packet in packets)
        {
            if ([packet isKindOfClass:[FMIceLinkRTCPReportPacket class]])
            {
                _reportsReceived++;
                
                FMIceLinkRTCPReportPacket *report = (FMIceLinkRTCPReportPacket *)packet;
                for (FMIceLinkRTCPReportBlock *block in report.reportBlocks)
                {
                    [FMLog debugWithMessage:[NSString stringWithFormat:@"Opus report: %0.2f %% packet loss (%d cumulative packets lost)", block.percentLost * 100, block.cumulativeNumberOfPacketsLost]];
                    if (block.percentLost > 0)
                    {
                        _losslessCount = 0;
                        _lossyCount++;
                        if (_lossyCount > 5 && _encoder.quality > 0.0)
                        {
                            _lossyCount = 0;
                            _encoder.quality = MAX(0.0, _encoder.quality - 0.1);
                            [FMLog infoWithMessage:[NSString stringWithFormat:@"Decreasing Opus encoder quality to %0.2f %%.", _encoder.quality * 100]];
                        }
                    }
                    else
                    {
                        _lossyCount = 0;
                        _losslessCount++;
                        if (_losslessCount > 5 && _encoder.quality < 1.0)
                        {
                            _losslessCount = 0;
                            _encoder.quality = MIN(1.0, _encoder.quality + 0.1);
                            [FMLog infoWithMessage:[NSString stringWithFormat:@"Increasing Opus encoder quality to %0.2f %%.", _encoder.quality * 100]];
                        }
                    }
                    
                    if (!_disableFEC && !_fecActive && _reportsReceived > _minimumReportsBeforeFEC)
                    {
                        if ((block.percentLost * 100) > _percentLossToTriggerFEC)
                        {
                            [FMLog infoWithMessage:@"Activating FEC for Opus audio stream."];
                            [_encoder activateFECWithPacketLossPercent:_percentLossToTriggerFEC];
                            _fecActive = true;
                        }
                    }
                }
            }
        }
    }
}

- (void)destroy
{
    if (_encoder)
    {
        [_encoder destroy];
        _encoder = nil;
    }
    
    if (_decoder)
    {
        [_decoder destroy];
        _decoder = nil;
    }
}

@end

File: Android.Conference.WebRTC/app/src/main/java/aopus/OpusCodec.java
package aopus;

import java.util.Locale;

import fm.*;
import fm.icelink.*;
import fm.icelink.webrtc.*;

public class OpusCodec extends AudioCodec {

    private BasicAudioPadep _padep;
    private Encoder _encoder;
    private Decoder _decoder;
    
    private OpusEchoCanceller _echoCanceller;
    public OpusEchoCanceller getEchoCanceller()
    {
        return _echoCanceller;
    }
    public void setEchoCanceller(OpusEchoCanceller echoCanceller)
    {
        _echoCanceller = echoCanceller;
    }
    
    private int _percentLossToTriggerFEC;
    public int getPercentLossToTriggerFEC()
    {
        return _percentLossToTriggerFEC;
    }
    public void setPercentLossToTriggerFEC(int percentLossToTriggerFEC)
    {
        _percentLossToTriggerFEC = percentLossToTriggerFEC;
    }
    
    private boolean _disableFEC;
    public boolean getDisableFEC()
    {
        return _disableFEC;
    }
    public void setDisableFEC(boolean disableFEC)
    {
        _disableFEC = disableFEC;
    }
    
    private boolean _fecActive;
    
    public OpusCodec()
    {
        this(null);
    }

    public OpusCodec(OpusEchoCanceller echoCanceller)
    {
        super(20);
        
        _disableFEC = false;
        _percentLossToTriggerFEC = 5;
        
        _echoCanceller = echoCanceller;
        _padep = new BasicAudioPadep();
    }

    /// <summary>
    /// Encodes a frame.
    /// </summary>
    /// <param name="frame">The frame.</param>
    /// <returns></returns>
    public byte[] encode(AudioBuffer frame)
    {
        if (_encoder == null)
        {
            _encoder = new Encoder(getClockRate(), getChannels(), getPacketTime());
            _encoder.setQuality(0.5);
            _encoder.setBitrate(64);
        }
        
        byte[] data; int index; int length;
        OpusEchoCanceller echoCanceller = getEchoCanceller();
        if (echoCanceller == null)
        {
            data = frame.getData();
            index = frame.getIndex();
            length = frame.getLength();
        }
        else
        {
            data = echoCanceller.capture(frame);
            index = 0;
            length = data.length;
        } 
        
        return _encoder.encode(data, index, length);
    }

    private int _currentRTPSequenceNumber = -1;
    private int _lastRTPSequenceNumber = -1;

    /// <summary>
    /// Decodes an encoded frame.
    /// </summary>
    /// <param name="encodedFrame">The encoded frame.</param>
    /// <returns></returns>
    public AudioBuffer decode(byte[] encodedFrame)
    {
        if (_decoder == null)
        {
            _decoder = new Decoder(getClockRate(), getChannels(), getPacketTime());
            LinkExtensions.getRemoteStream(getLink()).setDisablePLC(true);
        }

        if (_lastRTPSequenceNumber == -1)
        {
            _lastRTPSequenceNumber = _currentRTPSequenceNumber;
            return decodeNormal(encodedFrame);
        }
        else
        {
            int sequenceNumberDelta = RTPPacket.getSequenceNumberDelta(_currentRTPSequenceNumber, _lastRTPSequenceNumber);
            if (sequenceNumberDelta <= 0)
            {
                return null;
            }

            _lastRTPSequenceNumber = _currentRTPSequenceNumber;

            int missingPacketCount = sequenceNumberDelta - 1;
            AudioBuffer[] previousFrames = new AudioBuffer[missingPacketCount];

            int plcFrameCount = (missingPacketCount > 1) ? missingPacketCount - 1 : 0;
            if (plcFrameCount > 0)
            {
                Log.info(String.format(Locale.getDefault(), "Adding %d frames of loss concealment to incoming audio stream. Packet sequence violated.", plcFrameCount));
                for (int i = 0; i < plcFrameCount; i++)
                {
                    previousFrames[i] = decodePLC();
                }
            }

            int fecFrameCount = (missingPacketCount > 0) ? 1 : 0;
            if (fecFrameCount > 0)
            {
                AudioBuffer fecFrame = decodeFEC(encodedFrame);
                int fecFrameIndex = missingPacketCount - 1;
                if (fecFrame == null)
                {
                    previousFrames[fecFrameIndex] = decodePLC();
                }
                else
                {
                    previousFrames[fecFrameIndex] = fecFrame;
                }
            }
            
            AudioBuffer frame = decodeNormal(encodedFrame);
            frame.setPreviousBuffers(previousFrames);
            return frame;
        }
    }

    private AudioBuffer decodePLC()
    {
        return decode(null, false);
    }

    private AudioBuffer decodeFEC(byte[] encodedFrame)
    {
        return decode(encodedFrame, true);
    }

    private AudioBuffer decodeNormal(byte[] encodedFrame)
    {
        return decode(encodedFrame, false);
    }

    private AudioBuffer decode(byte[] encodedFrame, boolean fec)
    {
        byte[] data = _decoder.decode(encodedFrame, fec);
        if (data == null)
        {
            return null;
        }

        AudioBuffer frame = new AudioBuffer(data, 0, data.length);
        OpusEchoCanceller echoCanceller = getEchoCanceller();
        if (echoCanceller != null)
        {
            echoCanceller.render(getPeerId(), frame);
        }
        return frame;
    }

    /// <summary>
    /// Packetizes an encoded frame.
    /// </summary>
    /// <param name="encodedFrame">The encoded frame.</param>
    /// <returns></returns>
    public RTPPacket[] packetize(byte[] encodedFrame)
    {
        return _padep.packetize(encodedFrame, getClockRate(), getPacketTime(), getResetTimestamp());
    }

    /// <summary>
    /// Depacketizes a packet.
    /// </summary>
    /// <param name="packet">The packet.</param>
    /// <returns></returns>
    public byte[] depacketize(RTPPacket packet)
    {
        _currentRTPSequenceNumber = packet.getSequenceNumber();
        
        return _padep.depacketize(packet);
    }

    private int _lossyCount = 0;
    private int _losslessCount = 0;

    private int _minimumReportsBeforeFEC = 1;
    private long _reportsReceived = 0;

    /// <summary>
    /// Processes RTCP packets.
    /// </summary>
    /// <param name="packets">The packets to process.</param>
    public void processRTCP(RTCPPacket[] packets)
    {
        if (_encoder != null)
        {
            for (RTCPPacket packet : packets)
            {
                if (packet instanceof RTCPReportPacket)
                {
                    _reportsReceived++;

                    RTCPReportPacket report = (RTCPReportPacket)packet;
                    for (RTCPReportBlock block : report.getReportBlocks())
                    {
                        Log.debug(String.format(Locale.getDefault(), "Opus report: %.2f %% packet loss (%d cumulative packets lost)", block.getPercentLost() * 100, block.getCumulativeNumberOfPacketsLost()));
                        if (block.getPercentLost() > 0)
                        {
                            _losslessCount = 0;
                            _lossyCount++;
                            if (_lossyCount > 5 && _encoder.getQuality() > 0.0)
                            {
                                _lossyCount = 0;
                                _encoder.setQuality(MathAssistant.max(0.0, _encoder.getQuality() - 0.1));
                                Log.info(String.format(Locale.getDefault(), "Decreasing Opus encoder quality to %.2f %%.", _encoder.getQuality() * 100));
                            }
                        }
                        else
                        {
                            _lossyCount = 0;
                            _losslessCount++;
                            if (_losslessCount > 5 && _encoder.getQuality() < 1.0)
                            {
                                _losslessCount = 0;
                                _encoder.setQuality(MathAssistant.min(1.0, _encoder.getQuality() + 0.1));
                                Log.info(String.format(Locale.getDefault(), "Increasing Opus encoder quality to %.2f %%.", _encoder.getQuality() * 100));
                            }
                        }

                        if (!_disableFEC && !_fecActive && _reportsReceived > _minimumReportsBeforeFEC)
                        {
                            if ((block.getPercentLost() * 100) > _percentLossToTriggerFEC)
                            {
                                Log.info("Activating FEC for Opus audio stream.");
                                _encoder.activateFEC(_percentLossToTriggerFEC);
                                _fecActive = true;
                            }
                        }
                    }
                }
            }
        }
    }

    /// <summary>
    /// Destroys the codec.
    /// </summary>
    public void destroy()
    {
        if (_encoder != null)
        {
            _encoder.destroy();
            _encoder = null;
        }

        if (_decoder != null)
        {
            _decoder.destroy();
            _decoder = null;
        }
    }
}
